include/pybind11/pytypes.h

```diff
@@ -1137,11 +1152,18 @@ class int_ : public object {
     template <typename T,
               detail::enable_if_t<std::is_integral<T>::value, int> = 0>
     operator T() const {
-        return std::is_unsigned<T>::value
-            ? detail::as_unsigned<T>(m_ptr)
-            : sizeof(T) <= sizeof(long)
-              ? (T) PyLong_AsLong(m_ptr)
-              : (T) PYBIND11_LONG_AS_LONGLONG(m_ptr);
+        using py_type = detail::py_int_type_for<T>;
+        auto value = detail::as_integer<py_type>(m_ptr);
+        if (PyErr_Occurred())
+            throw error_already_set();
```

Comment:
Isn't this now throwing `OverflowError` instead of `RuntimeError`?

Reply:
Yes, this is what's currently making tests fail.